File: lib/features/compatibility/compatibility_service.dart
import 'package:uuid/uuid.dart';
import 'models.dart';
import 'questions_data.dart';

class CompatibilityService {
  static const Uuid _uuid = Uuid();

  /// Calculate compatibility between two users based on their responses
  static CompatibilityScore calculateCompatibility(
    CompatibilityResponse user1Response,
    CompatibilityResponse user2Response,
  ) {
    final categories = IslamicCompatibilityQuestions.getCategories();
    final Map<String, double> categoryScores = {};
    double totalWeightedScore = 0.0;

    for (final category in categories) {
      final double categoryScore = _calculateCategoryScore(
        category,
        user1Response.responses,
        user2Response.responses,
      );
      categoryScores[category.id] = categoryScore;
      totalWeightedScore += categoryScore * category.weight;
    }

    // Normalize to 0-100 scale
    final double overallScore = (totalWeightedScore * 100);

    return CompatibilityScore(
      userId1: user1Response.userId,
      userId2: user2Response.userId,
      overallScore: overallScore,
      categoryScores: categoryScores,
      calculatedAt: DateTime.now(),
      detailedBreakdown: _generateDetailedBreakdown(categoryScores, categories),
    );
  }

  /// Calculate score for a specific category
  static double _calculateCategoryScore(
    IslamicCompatibilityCategory category,
    Map<String, dynamic> user1Responses,
    Map<String, dynamic> user2Responses,
  ) {
    double totalScore = 0.0;
    int questionCount = 0;

    for (final question in category.questions) {
      final double questionScore = _calculateQuestionScore(
        question,
        user1Responses[question.id],
        user2Responses[question.id],
      );
      
      if (questionScore >= 0) { // Valid comparison
        totalScore += questionScore;
        questionCount++;
      }
    }

    return questionCount > 0 ? totalScore / questionCount : 0.0;
  }

  /// Calculate score for a single question
  static double _calculateQuestionScore(
    CompatibilityQuestion question,
    dynamic user1Answer,
    dynamic user2Answer,
  ) {
    if (user1Answer == null || user2Answer == null) {
      return -1.0; // Invalid comparison
    }

    switch (question.type) {
      case QuestionType.singleChoice:
      case QuestionType.yesNo:
        return _compareSingleChoice(question, user1Answer, user2Answer);
      
      case QuestionType.scale:
        return _compareScale(question, user1Answer, user2Answer);
      
      case QuestionType.multipleChoice:
        return _compareMultipleChoice(question, user1Answer, user2Answer);
    }
  }

  /// Compare single choice answers
  static double _compareSingleChoice(
    CompatibilityQuestion question,
    String user1Answer,
    String user2Answer,
  ) {
    final option1 = question.options.firstWhere((opt) => opt.id == user1Answer);
    final option2 = question.options.firstWhere((opt) => opt.id == user2Answer);
    
    // Direct value comparison
    if (option1.value == option2.value) {
      return 1.0;
    }
    
    // Calculate similarity based on value difference
    final double difference = (option1.value - option2.value).abs();
    return 1.0 - difference;
  }

  /// Compare scale answers
  static double _compareScale(
    CompatibilityQuestion question,
    String user1Answer,
    String user2Answer,
  ) {
    final option1 = question.options.firstWhere((opt) => opt.id == user1Answer);
    final option2 = question.options.firstWhere((opt) => opt.id == user2Answer);
    
    // Calculate similarity based on value difference
    final double difference = (option1.value - option2.value).abs();
    return 1.0 - difference;
  }

  /// Compare multiple choice answers
  static double _compareMultipleChoice(
    CompatibilityQuestion question,
    List<String> user1Answers,
    List<String> user2Answers,
  ) {
    if (user1Answers.isEmpty && user2Answers.isEmpty) {
      return 1.0;
    }
    
    // Calculate overlap
    final Set<String> set1 = user1Answers.toSet();
    final Set<String> set2 = user2Answers.toSet();
    final Set<String> intersection = set1.intersection(set2);
    final Set<String> union = set1.union(set2);
    
    if (union.isEmpty) {
      return 1.0;
    }
    
    return intersection.length / union.length;
  }

  /// Generate detailed breakdown for analysis
  static Map<String, dynamic> _generateDetailedBreakdown(
    Map<String, double> categoryScores,
    List<IslamicCompatibilityCategory> categories,
  ) {
    final Map<String, dynamic> breakdown = {};
    
    for (final category in categories) {
      final score = categoryScores[category.id] ?? 0.0;
      breakdown[category.id] = {
        'name': category.name,
        'score': score,
        'weight': category.weight,
        'weightedScore': score * category.weight,
        'description': _getCategoryDescription(score),
      };
    }
    
    return breakdown;
  }

  /// Get description for category score
  static String _getCategoryDescription(double score) {
    if (score >= 0.9) {
      return 'Excellent alignment in this area';
    } else if (score >= 0.8) {
      return 'Strong compatibility in this area';
    } else if (score >= 0.7) {
      return 'Good match in this area';
    } else if (score >= 0.6) {
      return 'Moderate compatibility - some discussion needed';
    } else if (score >= 0.5) {
      return 'Fair compatibility - may require compromise';
    } else {
      return 'Low compatibility - careful consideration needed';
    }
  }

  /// Generate personalized recommendations based on compatibility scores
  static List<String> generateRecommendations(CompatibilityScore score) {
    final List<String> recommendations = [];
    
    // Overall recommendations
    if (score.overallScore >= 80) {
      recommendations.add('You have excellent compatibility! Focus on building on your shared strengths.');
    } else if (score.overallScore >= 70) {
      recommendations.add('Good compatibility! Discuss minor differences to strengthen your connection.');
    } else if (score.overallScore >= 60) {
      recommendations.add('Moderate compatibility. Open communication about differences will be important.');
    } else {
      recommendations.add('Take time to understand your differences before making commitments.');
    }
    
    // Category-specific recommendations
    for (final categoryEntry in score.categoryScores.entries) {
      final categoryId = categoryEntry.key;
      final categoryScore = categoryEntry.value;
      
      if (categoryScore < 0.6) {
        recommendations.add(_getCategoryRecommendation(categoryId, categoryScore));
      }
    }
    
    return recommendations;
  }

  /// Get specific recommendation for a category
  static String _getCategoryRecommendation(String categoryId, double score) {
    switch (categoryId) {
      case 'religious_practice':
        return 'Discuss your religious practices and find common ground in daily worship.';
      case 'family_structure':
        return 'Have open conversations about family expectations and parenting approaches.';
      case 'cultural_values':
        return 'Explore each other\'s cultural backgrounds and find ways to honor both traditions.';
      case 'financial_management':
        return 'Create a financial plan that respects both your values regarding halal income and charity.';
      case 'education_career':
        return 'Discuss career goals and work-life balance expectations early in your relationship.';
      case 'social_life':
        return 'Find a balance between community involvement and personal time together.';
      case 'conflict_resolution':
        return 'Establish healthy communication patterns for resolving disagreements respectfully.';
      case 'life_goals':
        return 'Align on major life decisions including location and family planning.';
      default:
        return 'Take time to understand your differences in this important area.';
    }
  }

  /// Create a shareable compatibility report
  static CompatibilityReport createReport(
    CompatibilityScore scores,
    String userId1,
    String userId2, {
    List<FamilyFeedback>? familyFeedback,
  }) {
    return CompatibilityReport(
      id: _uuid.v4(),
      userId1: userId1,
      userId2: userId2,
      scores: scores,
      generatedAt: DateTime.now(),
      familyFeedback: familyFeedback,
    );
  }

  /// Add family feedback to a report
  static FamilyFeedback addFamilyFeedback({
    required String reportId,
    required String familyMemberName,
    required String relationship,
    required String feedback,
  }) {
    return FamilyFeedback(
      id: _uuid.v4(),
      reportId: reportId,
      familyMemberName: familyMemberName,
      relationship: relationship,
      feedback: feedback,
      createdAt: DateTime.now(),
    );
  }

  /// Validate that all required questions are answered
  static bool validateResponse(
    CompatibilityResponse response,
    List<IslamicCompatibilityCategory> categories,
  ) {
    for (final category in categories) {
      for (final question in category.questions) {
        if (question.isRequired && !response.responses.containsKey(question.id)) {
          return false;
        }
      }
    }
    return true;
  }

  /// Get unanswered questions for a response
  static List<CompatibilityQuestion> getUnansweredQuestions(
    CompatibilityResponse response,
    List<IslamicCompatibilityCategory> categories,
  ) {
    final List<CompatibilityQuestion> unanswered = [];
    
    for (final category in categories) {
      for (final question in category.questions) {
        if (question.isRequired && !response.responses.containsKey(question.id)) {
          unanswered.add(question);
        }
      }
    }
    
    return unanswered;
  }
}

File: native_ios/Sources/Features/Compatibility/CompatibilityService.swift
import Foundation
import Combine

#if os(iOS)

/// Service for calculating compatibility scores between users
@available(iOS 17, *)
@MainActor
public final class CompatibilityService: ObservableObject {
    
    // MARK: - Published Properties
    
    @Published public private(set) var userResponse: CompatibilityResponse?
    @Published public private(set) var responseState = UserResponseState()
    @Published public private(set) var currentScore: CompatibilityScore?
    @Published public private(set) var isLoading = false
    @Published public private(set) var error: Error?
    
    private let repository: CompatibilityRepository
    private let analytics: AnalyticsService
    
    // MARK: - Initialization
    
    public init(
        repository: CompatibilityRepository = CompatibilityRepository(),
        analytics: AnalyticsService = AnalyticsService.shared
    ) {
        self.repository = repository
        self.analytics = analytics
    }
    
    // MARK: - Response Management
    
    /// Save or update a user's response to a specific question
    public func saveResponse(questionId: String, value: ResponseValue) {
        responseState.responses[questionId] = value
        analytics.track(AnalyticsEvent(
            name: "compatibility_question_answered",
            parameters: [
                "question_id": questionId,
                "progress": "\(responseState.progressPercentage)"
            ]
        ))
    }
    
    /// Submit completed questionnaire
    public func submitQuestionnaire(userId: String) async throws {
        isLoading = true
        error = nil
        
        do {
            let response = CompatibilityResponse(
                userId: userId,
                responses: responseState.responses,
                completedAt: Date()
            )
            
            try await repository.saveResponse(response)
            userResponse = response
            
            analytics.track(AnalyticsEvent(
                name: "compatibility_questionnaire_completed",
                parameters: [
                    "user_id": userId,
                    "total_questions": "\(responseState.responses.count)"
                ]
            ))
            
            isLoading = false
        } catch {
            self.error = error
            isLoading = false
            throw error
        }
    }
    
    /// Load user's existing response
    public func loadUserResponse(userId: String) async {
        isLoading = true
        error = nil
        
        do {
            if let response = try await repository.fetchResponse(userId: userId) {
                userResponse = response
                responseState.responses = response.responses
            }
            isLoading = false
        } catch {
            self.error = error
            isLoading = false
        }
    }
    
    // MARK: - Compatibility Calculation
    
    /// Calculate compatibility between two users
    public func calculateCompatibility(
        user1Response: CompatibilityResponse,
        user2Response: CompatibilityResponse
    ) -> CompatibilityScore {
        let categories = IslamicCompatibilityQuestions.getCategories()
        var categoryScores: [String: Double] = [:]
        var totalWeightedScore: Double = 0.0
        
        for category in categories {
            let categoryScore = calculateCategoryScore(
                category: category,
                user1Responses: user1Response.responses,
                user2Responses: user2Response.responses
            )
            categoryScores[category.id] = categoryScore
            totalWeightedScore += categoryScore * category.weight
        }
        
        // Normalize to 0-100 scale
        let overallScore = totalWeightedScore * 100
        
        let score = CompatibilityScore(
            userId1: user1Response.userId,
            userId2: user2Response.userId,
            overallScore: overallScore,
            categoryScores: categoryScores,
            calculatedAt: Date(),
            detailedBreakdown: generateDetailedBreakdown(categoryScores: categoryScores, categories: categories)
        )
        
        analytics.track(AnalyticsEvent(
            name: "compatibility_score_calculated",
            parameters: [
                "user1_id": user1Response.userId,
                "user2_id": user2Response.userId,
                "overall_score": "\(overallScore)",
                "compatibility_level": score.compatibilityLevel
            ]
        ))
        
        return score
    }
    
    /// Calculate compatibility score for a specific category
    private func calculateCategoryScore(
        category: IslamicCompatibilityCategory,
        user1Responses: [String: ResponseValue],
        user2Responses: [String: ResponseValue]
    ) -> Double {
        var totalScore: Double = 0.0
        var questionCount = 0
        
        for question in category.questions {
            if let questionScore = calculateQuestionScore(
                question: question,
                user1Answer: user1Responses[question.id],
                user2Answer: user2Responses[question.id]
            ), questionScore >= 0 {
                totalScore += questionScore
                questionCount += 1
            }
        }
        
        return questionCount > 0 ? totalScore / Double(questionCount) : 0.0
    }
    
    /// Calculate score for a single question
    private func calculateQuestionScore(
        question: CompatibilityQuestion,
        user1Answer: ResponseValue?,
        user2Answer: ResponseValue?
    ) -> Double? {
        guard let answer1 = user1Answer, let answer2 = user2Answer else {
            return nil // Invalid comparison
        }
        
        switch question.type {
        case .singleChoice, .yesNo:
            return compareSingleChoice(question: question, answer1: answer1, answer2: answer2)
        case .scale:
            return compareScale(question: question, answer1: answer1, answer2: answer2)
        case .multipleChoice:
            return compareMultipleChoice(question: question, answer1: answer1, answer2: answer2)
        }
    }
    
    /// Compare single choice answers
    private func compareSingleChoice(
        question: CompatibilityQuestion,
        answer1: ResponseValue,
        answer2: ResponseValue
    ) -> Double {
        guard case .single(let value1) = answer1,
              case .single(let value2) = answer2,
              let option1 = question.options.first(where: { $0.id == value1 }),
              let option2 = question.options.first(where: { $0.id == value2 }) else {
            return 0.0
        }
        
        // Direct value comparison
        if option1.value == option2.value {
            return 1.0
        }
        
        // Calculate similarity based on value difference
        let difference = abs(option1.value - option2.value)
        return max(0.0, 1.0 - difference)
    }
    
    /// Compare scale answers
    private func compareScale(
        question: CompatibilityQuestion,
        answer1: ResponseValue,
        answer2: ResponseValue
    ) -> Double {
        guard case .single(let value1) = answer1,
              case .single(let value2) = answer2,
              let option1 = question.options.first(where: { $0.id == value1 }),
              let option2 = question.options.first(where: { $0.id == value2 }) else {
            return 0.0
        }
        
        // Calculate similarity based on value difference
        let difference = abs(option1.value - option2.value)
        return max(0.0, 1.0 - difference)
    }
    
    /// Compare multiple choice answers (Jaccard similarity)
    private func compareMultipleChoice(
        question: CompatibilityQuestion,
        answer1: ResponseValue,
        answer2: ResponseValue
    ) -> Double {
        guard case .multiple(let values1) = answer1,
              case .multiple(let values2) = answer2 else {
            return 0.0
        }
        
        if values1.isEmpty && values2.isEmpty {
            return 1.0
        }
        
        // Calculate Jaccard similarity (intersection / union)
        let set1 = Set(values1)
        let set2 = Set(values2)
        let intersection = set1.intersection(set2)
        let union = set1.union(set2)
        
        guard !union.isEmpty else { return 1.0 }
        
        return Double(intersection.count) / Double(union.count)
    }
    
    /// Generate detailed breakdown for analysis
    private func generateDetailedBreakdown(
        categoryScores: [String: Double],
        categories: [IslamicCompatibilityCategory]
    ) -> [String: Double] {
        var breakdown: [String: Double] = [:]
        
        for category in categories {
            if let score = categoryScores[category.id] {
                breakdown["\(category.id)_score"] = score
                breakdown["\(category.id)_weighted"] = score * category.weight
            }
        }
        
        return breakdown
    }
    
    // MARK: - Report Management
    
    /// Generate and save compatibility report
    public func generateReport(
        userId1: String,
        userId2: String
    ) async throws -> CompatibilityReport {
        isLoading = true
        error = nil
        
        do {
            guard let response1 = try await repository.fetchResponse(userId: userId1),
                  let response2 = try await repository.fetchResponse(userId: userId2) else {
                throw CompatibilityError.responsesNotFound
            }
            
            let score = calculateCompatibility(
                user1Response: response1,
                user2Response: response2
            )
            
            let report = CompatibilityReport(
                id: UUID().uuidString,
                userId1: userId1,
                userId2: userId2,
                scores: score,
                generatedAt: Date(),
                familyFeedback: nil,
                isShared: false
            )
            
            try await repository.saveReport(report)
            
            analytics.track(AnalyticsEvent(
                name: "compatibility_report_generated",
                parameters: [
                    "report_id": report.id,
                    "user1_id": userId1,
                    "user2_id": userId2,
                    "overall_score": "\(score.overallScore)"
                ]
            ))
            
            isLoading = false
            return report
        } catch {
            self.error = error
            isLoading = false
            throw error
        }
    }
    
    /// Fetch compatibility reports for a user
    public func fetchReports(userId: String) async throws -> [CompatibilityReport] {
        isLoading = true
        error = nil
        
        do {
            let reports = try await repository.fetchReports(userId: userId)
            
            analytics.track(AnalyticsEvent(
                name: "compatibility_reports_loaded",
                parameters: [
                    "user_id": userId,
                    "report_count": "\(reports.count)"
                ]
            ))
            
            isLoading = false
            return reports
        } catch {
            self.error = error
            isLoading = false
            throw error
        }
    }
    
    /// Share report with family member
    public func shareReport(reportId: String, withUserId: String) async throws {
        try await repository.shareReport(reportId: reportId, withUserId: withUserId)
        
        analytics.track(AnalyticsEvent(
            name: "compatibility_report_shared",
            parameters: [
                "report_id": reportId,
                "shared_with": withUserId
            ]
        ))
    }
    
    // MARK: - Progress Tracking
    
    /// Get current progress through questionnaire
    public var progress: Double {
        responseState.progressPercentage
    }
    
    /// Check if questionnaire is complete
    public var isComplete: Bool {
        responseState.isComplete
    }
    
    /// Reset questionnaire state
    public func resetQuestionnaire() {
        responseState = UserResponseState()
        analytics.track(event: "compatibility_questionnaire_reset")
    }
}

// MARK: - Errors

public enum CompatibilityError: LocalizedError {
    case responsesNotFound
    case invalidResponse
    case saveFailed
    
    public var errorDescription: String? {
        switch self {
        case .responsesNotFound:
            return "Could not find compatibility responses for one or both users"
        case .invalidResponse:
            return "The response data is invalid or incomplete"
        case .saveFailed:
            return "Failed to save compatibility data"
        }
    }
}

#endif
